public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx

```diff
@@ -0,0 +1,89 @@
+// Libraries
+import React, { memo } from 'react';
+import _ from 'lodash';
+
+// Types
+import { AbsoluteTimeRange, ExploreQueryFieldProps, ExploreMode } from '@grafana/data';
+import { LokiDatasource } from '../datasource';
+import { LokiQuery, LokiOptions } from '../types';
+import { LokiQueryField } from './LokiQueryField';
+import LokiExploreExtraField from './LokiExploreExtraField';
+
+type Props = ExploreQueryFieldProps<LokiDatasource, LokiQuery, LokiOptions>;
+
+export function LokiExploreQueryEditor(props: Props) {
+  const { query, data, datasource, exploreMode, history, onChange, onRunQuery } = props;
+
+  let absolute: AbsoluteTimeRange;
+  if (data && !_.isEmpty(data.request)) {
+    const { range } = data.request;
+
+    absolute = {
+      from: range.from.valueOf(),
+      to: range.to.valueOf(),
+    };
+  } else {
+    absolute = {
+      from: Date.now() - 10000,
+      to: Date.now(),
+    };
+  }
+
+  function onChangeQueryLimit(value: string) {
+    const { query, onChange } = props;
+    const nextQuery = { ...query, maxLines: preprocessMaxLines(value) };
+    onChange(nextQuery);
+  }
+
+  function preprocessMaxLines(value: string): number {
+    if (value.length === 0) {
+      // empty input - falls back to dataSource.maxLines limit
+      return NaN;
+    } else if (value.length > 0 && (isNaN(+value) || +value < 0)) {
+      // input with at least 1 character and that is either incorrect (value in the input field is not a number) or negative
+      // falls back to the limit of 0 lines
+      return 0;
+    } else {
+      // default case - correct input
+      return +value;
+    }
+  }
+
+  function onMaxLinesChange(e: React.SyntheticEvent<HTMLInputElement>) {
+    if (e.currentTarget.value.length === 0 || parseInt(e.currentTarget.value, 10) !== query.maxLines) {
```

Comment:
Can't we just check if `query.maxLines` is different from `preprocessMaxLines (e.currentTarget.value)`? Seems like that would simplify it and I assume that's what we want, save the value if it differs from upstream right?

Reply:
The only issue with getting over this now is that we still get those NaNs on empty input coming from `preprocessMaxLines()` due to our *semi-elegant* workaround and it can mess up with input values that are being passed down to the component. It's going to change with an introduction of new forms as they'll cover the validation part. I'll address that in the next PR about this.